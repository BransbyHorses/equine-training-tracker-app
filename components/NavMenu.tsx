import * as React from "react";
import { signOut, useSession } from "next-auth/react";
import { useRouter } from "next/router";

import Drawer from "@mui/material/Drawer";
import List from "@mui/material/List";
import Divider from "@mui/material/Divider";
import IconButton from "@mui/material/IconButton";
import ChevronLeftIcon from "@mui/icons-material/ChevronLeft";
import ChevronRightIcon from "@mui/icons-material/ChevronRight";
import ListItem from "@mui/material/ListItem";
import ListItemButton from "@mui/material/ListItemButton";
import ListItemIcon from "@mui/material/ListItemIcon";
import ListItemText from "@mui/material/ListItemText";
import { Button } from "@mui/material";
import HomeIcon from "@mui/icons-material/Home";
import { styled, useTheme } from "@mui/material/styles";
import AdminPanelSettingsIcon from "@mui/icons-material/AdminPanelSettings";

const DrawerHeader = styled("div")(({ theme }) => ({
	display: "flex",
	alignItems: "center",
	padding: theme.spacing(0, 1),
	...theme.mixins.toolbar,
	justifyContent: "flex-start",
}));

interface NavMenuProps {
	drawerWidth: number;
	open: boolean;
	collapse: () => void;
}

const NavMenu = ({ drawerWidth, open, collapse }: NavMenuProps) => {
	const theme = useTheme();
	const router = useRouter();
	const session = useSession();
	return (
		<Drawer
			sx={{
				width: drawerWidth,
				flexShrink: 0,
				"& .MuiDrawer-paper": {
					width: drawerWidth,
				},
			}}
			variant="persistent"
			anchor="right"
			open={open}
		>
			<DrawerHeader>
				<IconButton onClick={collapse}>
					{theme.direction === "rtl" ? (
						<ChevronLeftIcon />
					) : (
						<ChevronRightIcon />
					)}
				</IconButton>
			</DrawerHeader>
			<Divider />
			<List>
				<ListItem disablePadding>
					<ListItemButton
						onClick={() => router.push(`${window.location.origin}/`)}
					>
						<ListItemIcon>
							<HomeIcon />
						</ListItemIcon>
						<ListItemText primary="Home" />
					</ListItemButton>
				</ListItem>
				{session.role === "ADMIN" && (
					<ListItem disablePadding>
						<ListItemButton
							onClick={() => router.push(`${window.location.origin}/`)}
						>
							<ListItemIcon>
								<AdminPanelSettingsIcon />
							</ListItemIcon>
							<ListItemText primary="Admin" />
						</ListItemButton>
					</ListItem>
				)}
			</List>
			<Divider />
			<List>
				<ListItem disablePadding>
					<ListItemButton>
						<Button
							color="primary"
							variant="contained"
							onClick={signOut}
							sx={{ width: "100%" }}
						>
							Sign Out
						</Button>
					</ListItemButton>
				</ListItem>
			</List>
		</Drawer>
	);
};

export default NavMenu;
