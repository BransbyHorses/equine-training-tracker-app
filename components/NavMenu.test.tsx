import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { fireEvent, render, screen } from "@testing-library/react";
import NavMenu, { NavMenuProps } from "./NavMenu";
import { useSession } from "next-auth/react";
import { createMockRouter } from "../utils/test-utils/mockNextRouter";
import { RouterContext } from "next/dist/shared/lib/router-context";

jest.mock("next-auth/react", () => ({
	__esModule: true,
	...jest.requireActual("next-auth/react"),
	useSession: jest.fn(),
}));

const props: NavMenuProps = {
	drawerWidth: 255,
	open: true,
	collapse: jest.fn(),
	signOut: jest.fn(),
};

const mockSession = {
	expires: new Date(Date.now() + 2 * 86400).toISOString(),
	user: { name: "Test", email: undefined, image: undefined },
};

describe("NavMenu", () => {
	it("renders with minimum props for admin user", () => {
		(useSession as jest.Mock).mockImplementation(() => {
			return {
				data: {
					...mockSession,
					role: "ADMIN",
				},
			};
		});
		const tree = renderer.create(<NavMenu {...props} />);
		expect(tree).toMatchSnapshot();
	});

	it("renders with minimum props for non-admin user", () => {
		(useSession as jest.Mock).mockImplementation(() => {
			return {
				data: {
					...mockSession,
					role: "",
				},
			};
		});
		const tree = renderer.create(<NavMenu {...props} />);
		expect(tree).toMatchSnapshot();
	});

	it("displays link to admin dashboard when user is an admin", () => {
		(useSession as jest.Mock).mockImplementation(() => {
			return {
				data: {
					...mockSession,
					role: "ADMIN",
				},
			};
		});
		render(<NavMenu {...props} />);
		expect(screen.getByText("Admin")).toBeInTheDocument();
	});

	it("hides link to admin dashboard when user is not an admin", () => {
		(useSession as jest.Mock).mockImplementation(() => {
			return {
				data: {
					...mockSession,
					role: "",
				},
			};
		});
		render(<NavMenu {...props} />);
		expect(screen.queryByText("Admin")).toBeNull();
	});

	it("displays sign out button when a user is logged in", () => {
		(useSession as jest.Mock).mockImplementation(() => {
			return {
				data: {
					...mockSession,
					role: "",
				},
			};
		});
		render(<NavMenu {...props} />);
		expect(
			screen.getByRole("button", { name: "Sign Out" })
		).toBeInTheDocument();
	});

	it("will collapse when user clicks close icon", () => {
		(useSession as jest.Mock).mockImplementation(() => {
			return {
				data: {
					...mockSession,
					role: "",
				},
			};
		});
		render(<NavMenu {...props} />);
		fireEvent.click(screen.getByTestId("ChevronRightIcon"));
		expect(props.collapse).toHaveBeenCalled();
	});

	it("will navigate to admin dashboard", () => {
		(useSession as jest.Mock).mockImplementation(() => {
			return {
				data: {
					...mockSession,
					role: "",
				},
			};
		});
		const router = createMockRouter({});
		render(
			<RouterContext.Provider value={router}>
				<NavMenu {...props} />
			</RouterContext.Provider>
		);
		fireEvent.click(screen.getByText("Home"));
		expect(router.push).toHaveBeenCalledWith("/");
	});

	it("will navigate to index page", () => {
		(useSession as jest.Mock).mockImplementation(() => {
			return {
				data: {
					...mockSession,
					role: "ADMIN",
				},
			};
		});
		const router = createMockRouter({});
		render(
			<RouterContext.Provider value={router}>
				<NavMenu {...props} />
			</RouterContext.Provider>
		);
		fireEvent.click(screen.getByText("Admin"));
		expect(router.push).toHaveBeenCalledWith("/admin");
	});
});
