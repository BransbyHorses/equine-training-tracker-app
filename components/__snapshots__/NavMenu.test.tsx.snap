// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavMenu renders with minimum props for admin user 1`] = `
<div
  className="MuiDrawer-root MuiDrawer-docked css-kyuk4x-MuiDrawer-docked"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorRight MuiDrawer-paperAnchorDockedRight css-18sg6k4-MuiPaper-root-MuiDrawer-paper"
    style={
      {
        "visibility": undefined,
      }
    }
  >
    <div
      className="css-fiov70"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronRightIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
      </button>
    </div>
    <hr
      className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <ul
      className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="HomeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Home
            </span>
          </div>
        </div>
      </li>
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="AdminPanelSettingsIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M17 11c.34 0 .67.04 1 .09V6.27L10.5 3 3 6.27v4.91c0 4.54 3.2 8.79 7.5 9.82.55-.13 1.08-.32 1.6-.55-.69-.98-1.1-2.17-1.1-3.45 0-3.31 2.69-6 6-6z"
              />
              <path
                d="M17 13c-2.21 0-4 1.79-4 4s1.79 4 4 4 4-1.79 4-4-1.79-4-4-4zm0 1.38c.62 0 1.12.51 1.12 1.12s-.51 1.12-1.12 1.12-1.12-.51-1.12-1.12.5-1.12 1.12-1.12zm0 5.37c-.93 0-1.74-.46-2.24-1.17.05-.72 1.51-1.08 2.24-1.08s2.19.36 2.24 1.08c-.5.71-1.31 1.17-2.24 1.17z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Admin
            </span>
          </div>
        </div>
      </li>
    </ul>
    <hr
      className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <ul
      className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
        disabled={false}
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[MockFunction]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Sign Out
        </button>
      </li>
    </ul>
  </div>
</div>
`;

exports[`NavMenu renders with minimum props for non-admin user 1`] = `
<div
  className="MuiDrawer-root MuiDrawer-docked css-kyuk4x-MuiDrawer-docked"
>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorRight MuiDrawer-paperAnchorDockedRight css-18sg6k4-MuiPaper-root-MuiDrawer-paper"
    style={
      {
        "visibility": undefined,
      }
    }
  >
    <div
      className="css-fiov70"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[MockFunction]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ChevronRightIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
          />
        </svg>
        <span
          className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <hr
      className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <ul
      className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters css-1nxmd3h-MuiListItem-root"
        disabled={false}
      >
        <div
          className="MuiButtonBase-root MuiListItemButton-root MuiListItemButton-gutters css-16ac5r2-MuiButtonBase-root-MuiListItemButton-root"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <div
            className="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="HomeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
              />
            </svg>
          </div>
          <div
            className="MuiListItemText-root css-tlelie-MuiListItemText-root"
          >
            <span
              className="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
            >
              Home
            </span>
          </div>
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </div>
      </li>
    </ul>
    <hr
      className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
    <ul
      className="MuiList-root MuiList-padding css-h4y409-MuiList-root"
    >
      <li
        className="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-1p823my-MuiListItem-root"
        disabled={false}
      >
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
          disabled={false}
          onBlur={[Function]}
          onClick={[MockFunction]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          Sign Out
          <span
            className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </li>
    </ul>
  </div>
</div>
`;
